File: src/tauri/events.ts
/*
 * Copyright 2025 Ruslan Curbanov
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { listen } from "@tauri-apps/api/event";
import { DaemonEvent } from "../types/daemon";
import { ViewStateEvent } from "../types/view-state";
import { EphemeralSessionEvent } from "../types/session";
import { ProjectEvent } from "../types/project";
import { ProxyEvent } from "../types/proxy";
import { ApiClientEvent } from "../types/api-client";

export type DevWebPenEvent =
  | DaemonEvent
  | ViewStateEvent
  | EphemeralSessionEvent
  | ProjectEvent
  | ProxyEvent
  | ApiClientEvent;

let listeners: Record<DevWebPenEvent, Set<(payload: any) => void>> = {
  "devwebpen://view-state-changed": new Set(),
  "devwebpen://daemon-state-changed": new Set(),
  "devwebpen://daemon-error": new Set(),
  "devwebpen://ephemeral-session-changed": new Set(),
  "devwebpen://project-changed": new Set(),
  "devwebpen://proxy-changed": new Set(),
  "devwebpen://api-client-changed": new Set(),
};

interface Subscription {
  event: DevWebPenEvent;
  unsubscribe: () => void;
}

export function subscribe(
  event: DevWebPenEvent,
  callback: (payload: any) => void
): Subscription {
  listeners[event].add(callback);

  return {
    event,
    unsubscribe: () => listeners[event].delete(callback),
  };
}

// register tauri event listeners
(Object.keys(listeners) as DevWebPenEvent[]).forEach((event) => {
  listen(event, ({ payload }) => {
    listeners[event].forEach((callback) => callback(payload));
  });
});

export default DevWebPenEvent;

File: src-tauri/src/events.rs
// Copyright 2025 Ruslan Curbanov
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use tauri::{AppHandle, Emitter};

use crate::{
    api_client::ApiClientInner,
    app_state::{project::Project, session::Session},
    daemon::DaemonState,
    proxy::ProxyInner,
    view::PartialViewState,
};

pub enum DevWebPenEvent {
    ViewStateChanged(PartialViewState),
    DaemonStateChanged(DaemonState),
    DaemonError(Option<String>),
    EphemeralSessionChanged(Option<Session>),
    ProjectChanged(Option<Project>),
    ProxyChanged(ProxyInner),
    ApiClientChanged(ApiClientInner),
}

impl DevWebPenEvent {
    fn to_string(&self) -> &'static str {
        match self {
            DevWebPenEvent::ViewStateChanged(_) => "devwebpen://view-state-changed",
            DevWebPenEvent::DaemonStateChanged(_) => "devwebpen://daemon-state-changed",
            DevWebPenEvent::DaemonError(_) => "devwebpen://daemon-error",
            DevWebPenEvent::EphemeralSessionChanged(_) => "devwebpen://ephemeral-session-changed",
            DevWebPenEvent::ProjectChanged(_) => "devwebpen://project-changed",
            DevWebPenEvent::ProxyChanged(_) => "devwebpen://proxy-changed",
            DevWebPenEvent::ApiClientChanged(_) => "devwebpen://api-client-changed",
        }
    }
}

pub fn emit_event(app_handle: &AppHandle, event: DevWebPenEvent) -> Result<(), tauri::Error> {
    let event_name = event.to_string();
    match event {
        DevWebPenEvent::ViewStateChanged(payload) => app_handle.emit(event_name, payload),
        DevWebPenEvent::DaemonStateChanged(payload) => app_handle.emit(event_name, payload),
        DevWebPenEvent::DaemonError(payload) => app_handle.emit(event_name, payload),
        DevWebPenEvent::EphemeralSessionChanged(payload) => app_handle.emit(event_name, payload),
        DevWebPenEvent::ProjectChanged(payload) => app_handle.emit(event_name, payload),
        DevWebPenEvent::ProxyChanged(payload) => app_handle.emit(event_name, payload),
        DevWebPenEvent::ApiClientChanged(payload) => app_handle.emit(event_name, payload),
    }
}
